File: web/frontend/lib/watchlist.ts
"use client";
import { useCallback, useEffect, useRef, useState } from "react";
import type { StockItem } from "./types";
import {
  getWatchlist,
  createGroup as apiCreateGroup,
  deleteGroup as apiDeleteGroup,
  renameGroup as apiRenameGroup,
  addStock as apiAddStock,
  removeStock as apiRemoveStock,
  toggleStock as apiToggleStock,
  importWatchlist,
} from "./api";

export interface WatchGroup {
  id: string;
  name: string;
  stocks: StockItem[];
}

export interface WatchlistApi {
  groups: WatchGroup[];
  ready: boolean;
  addGroup: (name: string) => Promise<WatchGroup>;
  removeGroup: (id: string) => Promise<void>;
  renameGroup: (id: string, name: string) => Promise<void>;
  addStock: (groupId: string, stock: StockItem) => Promise<void>;
  removeStock: (groupId: string, code: string) => Promise<void>;
  toggleStock: (groupId: string, stock: StockItem) => Promise<void>;
  isInGroup: (groupId: string, code: string) => boolean;
}

// 旧版浏览器 localStorage 键（迁移后清除）
const LEGACY_KEY = "stockmarket_watchlist_v1";

// 判断后端是否仍处于初始播种态（仅一个"我的自选"且为空），此时才允许从旧 localStorage 迁移
function isInitialState(groups: WatchGroup[]): boolean {
  return (
    groups.length === 1 &&
    groups[0].name === "我的自选" &&
    (!groups[0].stocks || groups[0].stocks.length === 0)
  );
}

export function useWatchlist(): WatchlistApi {
  const [groups, setGroups] = useState<WatchGroup[]>([]);
  const [ready, setReady] = useState(false);
  const migrated = useRef(false);

  // 启动时从后端（web/backend/data/watchlist.json）拉取分组；首次使用做一次旧数据迁移
  useEffect(() => {
    let alive = true;
    (async () => {
      try {
        const g = await getWatchlist();
        if (!alive) return;
        // 一次性迁移：仅当后端为空态且浏览器有旧数据时，导入并清理旧键
        if (!migrated.current && isInitialState(g)) {
          const legacy = (() => {
            try {
              const raw = localStorage.getItem(LEGACY_KEY);
              return raw ? (JSON.parse(raw) as WatchGroup[]) : null;
            } catch {
              return null;
            }
          })();
          if (legacy && Array.isArray(legacy) && legacy.length > 0) {
            const imported = await importWatchlist(legacy);
            if (alive) setGroups(imported);
            migrated.current = true;
            try {
              localStorage.removeItem(LEGACY_KEY);
            } catch {
              /* ignore */
            }
            return;
          }
        }
        setGroups(g);
      } catch {
        /* 后端不可用时保持空分组 */
      } finally {
        if (alive) setReady(true);
      }
    })();
    return () => {
      alive = false;
    };
  }, []);

  const addGroup = useCallback(async (name: string): Promise<WatchGroup> => {
    const res = await apiCreateGroup(name);
    setGroups(res.groups);
    return res.group;
  }, []);

  const removeGroup = useCallback(async (id: string) => {
    setGroups(await apiDeleteGroup(id));
  }, []);

  const renameGroup = useCallback(async (id: string, name: string) => {
    setGroups(await apiRenameGroup(id, name));
  }, []);

  const addStock = useCallback(async (groupId: string, stock: StockItem) => {
    setGroups(await apiAddStock(groupId, stock.code, stock.name));
  }, []);

  const removeStock = useCallback(async (groupId: string, code: string) => {
    setGroups(await apiRemoveStock(groupId, code));
  }, []);

  const toggleStock = useCallback(async (groupId: string, stock: StockItem) => {
    setGroups(await apiToggleStock(groupId, stock.code, stock.name));
  }, []);

  const isInGroup = useCallback(
    (groupId: string, code: string) =>
      groups.some((g) => g.id === groupId && g.stocks.some((s) => s.code === code)),
    [groups]
  );

  return { groups, ready, addGroup, removeGroup, renameGroup, addStock, removeStock, toggleStock, isInGroup };
}

File: web/backend/routers/watchlist.py
"""自选股 / 分组管理路由（数据持久化于 web/backend/data/watchlist.json）。

所有写操作返回更新后的完整分组列表，前端直接以服务端为准覆盖本地状态，避免二次 GET。
"""
from typing import Any, Dict, List

from fastapi import APIRouter
from pydantic import BaseModel, Field

from ..config import get_settings
from ..services import watchlist as svc

router = APIRouter()


class StockIn(BaseModel):
    code: str
    name: str = ""


class GroupName(BaseModel):
    name: str = ""


class ImportBody(BaseModel):
    groups: List[Dict[str, Any]] = Field(default_factory=list)


def _dir() -> str:
    return get_settings().backend_data_dir


@router.get("/watchlist")
def get_watchlist() -> List[Dict[str, Any]]:
    """获取全部分组。"""
    return svc.get_groups(_dir())


@router.post("/watchlist/groups")
def create_group(body: GroupName) -> Dict[str, Any]:
    """新建分组，返回新建对象 + 完整列表。"""
    g = svc.add_group(_dir(), body.name)
    return {"group": g, "groups": svc.get_groups(_dir())}


@router.delete("/watchlist/groups/{group_id}")
def delete_group(group_id: str) -> List[Dict[str, Any]]:
    """删除分组（保留至少一个）。"""
    return svc.remove_group(_dir(), group_id)


@router.patch("/watchlist/groups/{group_id}")
def rename_group(group_id: str, body: GroupName) -> List[Dict[str, Any]]:
    """重命名分组。"""
    return svc.rename_group(_dir(), group_id, body.name)


@router.post("/watchlist/groups/{group_id}/stocks")
def add_stock(group_id: str, body: StockIn) -> List[Dict[str, Any]]:
    """向分组添加一只股票（去重）。"""
    return svc.add_stock(_dir(), group_id, body.code, body.name)


@router.delete("/watchlist/groups/{group_id}/stocks/{code}")
def remove_stock(group_id: str, code: str) -> List[Dict[str, Any]]:
    """从分组移除一只股票。"""
    return svc.remove_stock(_dir(), group_id, code)


@router.post("/watchlist/groups/{group_id}/toggle")
def toggle_stock(group_id: str, body: StockIn) -> List[Dict[str, Any]]:
    """在分组内切换一只股票的归属（有则删、无则加）。"""
    return svc.toggle_stock(_dir(), group_id, body.code, body.name)


@router.post("/watchlist/import")
def import_watchlist(body: ImportBody) -> List[Dict[str, Any]]:
    """整体替换分组（用于从浏览器 localStorage 一次性迁移）。"""
    return svc.import_groups(_dir(), body.groups)
